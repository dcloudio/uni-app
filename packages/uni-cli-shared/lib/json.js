const fs = require('fs')
const path = require('path')
const stripJsonComments = require('strip-json-comments')
const uniI18n = require('@dcloudio/uni-cli-i18n')

function parseJson (content, preprocess = false) {
  if (typeof content === 'string') {
    if (preprocess) {
      const preprocessor = require('@dcloudio/vue-cli-plugin-uni/packages/webpack-preprocess-loader/preprocess')
      const {
        jsPreprocessOptions
      } = require('./platform')
      content = preprocessor.preprocess(content, jsPreprocessOptions.context, {
        type: jsPreprocessOptions.type
      })
    }

    try {
      content = JSON.parse(stripJsonComments(content))
    } catch (e) {
      throw new Error('uni-app-compiler: ' + e.message)
    }
  }

  content = JSON.stringify(content)
    .replace(/\u2028/g, '\\u2028')
    .replace(/\u2029/g, '\\u2029')

  return JSON.parse(content)
}

function getJson (jsonFileName, preprocess = false) {
  const jsonFilePath = path.resolve(process.env.UNI_INPUT_DIR, jsonFileName)
  if (!fs.existsSync(jsonFilePath)) {
    throw new Error(jsonFilePath + ' ' + uniI18n.__('cliShared.doesNotExist'))
  }
  try {
    return parseJson(fs.readFileSync(jsonFilePath, 'utf8'), preprocess)
  } catch (e) {
    console.error(jsonFileName + uniI18n.__('cliShared.parsingFailed'))
  }
}

module.exports = {
  getJson,
  parseJson
}
