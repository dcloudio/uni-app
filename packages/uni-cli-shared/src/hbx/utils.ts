import path from 'path'
import { once } from '@dcloudio/uni-shared'

export const isInHBuilderX = once(() => {
  // 自动化测试传入了 HX_APP_ROOT(其实就是UNI_HBUILDERX_PLUGINS)
  if (process.env.HX_APP_ROOT) {
    process.env.UNI_HBUILDERX_PLUGINS = process.env.HX_APP_ROOT + '/plugins'
    return true
  }
  try {
    const { name } = require(path.resolve(
      process.cwd(),
      '../about/package.json'
    ))
    if (name === 'about') {
      process.env.UNI_HBUILDERX_PLUGINS = path.resolve(process.cwd(), '..')
      return true
    }
  } catch (e) {
    // console.error(e)
  }
  return false
})
