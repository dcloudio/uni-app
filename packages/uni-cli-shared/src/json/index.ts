export * from './mp'
export * from './app'
export * from './json'
export * from './pages'
export * from './manifest'
export * from './theme'
export {
  normalizeUniAppXAppPagesJson,
  normalizeUniAppXAppConfig,
  checkPagesJson,
  parseUniXFlexDirection,
  parseUniXSplashScreen,
  parseUniXUniStatistics,
} from './uni-x'
