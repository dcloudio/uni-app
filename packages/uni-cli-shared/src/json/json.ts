import { parse } from 'jsonc-parser'
import { preJson } from '../preprocess'

export function parseJson<T = any>(jsonStr: string, shouldPre: boolean = false): T {
  return parse(shouldPre ? preJson(jsonStr) : jsonStr);
}
