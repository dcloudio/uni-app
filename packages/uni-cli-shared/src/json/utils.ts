import fs from 'fs'
import path from 'path'
import { isString } from '@vue/shared'
import type { CompilerError } from '@vue/compiler-core'
import {
  type Node,
  type ParseError,
  parseTree,
  printParseErrorCode,
} from 'jsonc-parser'

import { normalizePath } from '../utils'

import { offsetToLineColumn } from '../vite/plugins/vitejs/utils'

interface CheckPagesJsonError extends CompilerError {
  offsetStart: number
  offsetEnd: number
}

export function checkPagesJson(jsonStr: string, inputDir: string) {
  if (!inputDir) {
    return false
  }
  const errors: ParseError[] = []
  const root = parseTree(jsonStr, errors)
  if (!root) {
    if (errors.length) {
      for (const error of errors) {
        const { line, column } = offsetToLineColumn(jsonStr, error.offset)
        throw {
          name: 'SyntaxError',
          code: error.error,
          message: printParseErrorCode(error.error),
          loc: {
            start: { line, column },
          },
          offsetStart: error.offset,
          offsetEnd: error.offset + error.length,
        } as CheckPagesJsonError
      }
    }
    return false
  }
  const pagePathNodes = walkNodes(findRootNode(root, ['pages']))
  findRootNode(root, ['subPackages', 'subpackages']).forEach((node) => {
    const subPackageRoot = findSubPackageRoot(node)
    if (subPackageRoot) {
      findRootNode(node, ['pages']).forEach((subNode) => {
        walkNodes(subNode.children ?? []).forEach((node) => {
          pagePathNodes.push({
            ...node,
            value: normalizePath(path.join(subPackageRoot, node.value)),
          })
        })
      })
    }
  })

  if (pagePathNodes.length) {
    for (const node of pagePathNodes) {
      const pagePath: string = node.value ?? ''
      if (pageExistsWithCaseSync(path.join(inputDir, pagePath))) {
        continue
      }
      const { line, column } = offsetToLineColumn(jsonStr, node.offset)
      throw {
        name: 'CompilerError',
        code: 'CompilerError',
        message: `The page path "${pagePath}" does not exist`,
        loc: {
          start: { line, column },
        },
        offsetStart: node.offset,
        offsetEnd: node.offset + node.length,
      } as unknown as CheckPagesJsonError
    }
  }
  return true
}

function pageExistsWithCaseSync(pagePath: string) {
  try {
    const files = fs.readdirSync(path.dirname(pagePath))
    const basename = path.basename(pagePath)
    const uvuePage = basename + '.uvue'
    const vuePage = basename + '.vue'
    return files.some((file) => file === uvuePage || file === vuePage)
  } catch (e) {
    return false
  }
}

function findSubPackageRoot(node: Node) {
  const child = node.children?.find(
    (child) =>
      child.type === 'property' &&
      child.children &&
      child.children.find(
        (child) => child.type === 'string' && child.value === 'root'
      )
  )
  if (child && child.children?.length === 2) {
    return child.children[1].value
  }
  return ''
}

function findRootNode(node: Node, property: string[]) {
  const { type, children } = node
  if (type === 'object' && children) {
    const child = children.find(
      (child) =>
        child.type === 'property' &&
        child.children &&
        child.children.find(
          (child) => child.type === 'string' && property.includes(child.value)
        )
    )
    if (child) {
      const node = child.children!.find((child) => child.type === 'array')
      return node?.children ?? []
    }
  }
  return []
}

function walkNodes(node: Node[]) {
  const pagePathNodes: Node[] = []
  node.forEach((node) => walkNode(node, pagePathNodes))
  return pagePathNodes
}

function walkNode(node: Node, pagePathNodes: Node[]) {
  const { type, children } = node
  if (type === 'property' && children && children.length === 2) {
    const maybePagePathNode = children[0]
    const maybePagePathValueNode = children[1]
    if (
      maybePagePathNode.type === 'string' &&
      maybePagePathNode.value === 'path' &&
      maybePagePathValueNode.type === 'string' &&
      isString(maybePagePathValueNode.value)
    ) {
      pagePathNodes.push(maybePagePathValueNode)
    }
  }
  if (children) {
    children.forEach((node) => walkNode(node, pagePathNodes))
  }
}
