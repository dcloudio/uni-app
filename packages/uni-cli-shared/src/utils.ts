import fs from 'fs'
import os from 'os'
import path from 'path'
import colors from 'picocolors'
import { camelize, capitalize } from '@vue/shared'
export { default as hash } from 'hash-sum'
import {
  EXTNAME_TS_RE,
  PAGE_EXTNAME,
  PAGE_EXTNAME_APP,
  X_PAGE_EXTNAME,
  X_PAGE_EXTNAME_APP,
} from './constants'

import {
  type ElementNode,
  NodeTypes,
  type RootNode,
  type TemplateChildNode,
} from '@vue/compiler-core'
import type { ParserPlugin } from '@babel/parser'
import { getPlatformDir } from './platform'
import { isInHBuilderX } from './hbx'
import { parseManifestJsonOnce } from './json'

// 专为 uts.ts 服务
export { camelize, capitalize, isArray } from '@vue/shared'

// export let isRunningWithYarnPnp: boolean
// try {
//   isRunningWithYarnPnp = Boolean(require('pnpapi'))
// } catch {}

export const isWindows = os.platform() === 'win32'
export function normalizePath(id: string): string {
  return isWindows ? id.replace(/\\/g, '/') : id
}

export function checkElementNodeTag(
  node: RootNode | TemplateChildNode | null | undefined,
  tag: string
): node is ElementNode {
  return !!node && node.type === NodeTypes.ELEMENT && node.tag === tag
}

/**
 * 根据 path 返回合法 js 变量
 * @param str pages.json.page.path
 * @returns
 */
export function normalizeIdentifier(str: string) {
  let _str = str.replace(/[^a-zA-Z0-9]+/g, '-')
  _str = capitalize(camelize(_str))
  // 不允许数字开头，补充 _
  if (/^\d/.test(_str)) {
    _str = '_' + _str
  }
  return _str
}

export function normalizePagePath(pagePath: string, platform: UniApp.PLATFORM) {
  const absolutePagePath = path.resolve(process.env.UNI_INPUT_DIR, pagePath)
  const isX = process.env.UNI_APP_X === 'true'
  let extensions = isX ? X_PAGE_EXTNAME : PAGE_EXTNAME
  if (platform === 'app') {
    extensions = isX ? X_PAGE_EXTNAME_APP : PAGE_EXTNAME_APP
  }
  for (let i = 0; i < extensions.length; i++) {
    const extname = extensions[i]
    if (fs.existsSync(absolutePagePath + extname)) {
      return pagePath + extname
    }
  }
  console.error(`${pagePath} not found`)
}

export function removeExt(str: string) {
  return str.split('?')[0].replace(/\.\w+$/g, '')
}

const NODE_MODULES_REGEX = /(\.\.\/)?node_modules/g

export function normalizeNodeModules(str: string) {
  str = normalizePath(str).replace(NODE_MODULES_REGEX, 'node-modules')
  // HBuilderX 内置模块路径转换
  str = str.replace(
    /.*\/plugins\/uniapp-cli-vite\/node[-_]modules/,
    'node-modules'
  )
  if (!isInHBuilderX()) {
    // 内部测试
    if (str.includes('uni-app-next/packages/')) {
      str = str.replace(
        /.*\/uni-app-next\/packages\//,
        'node-modules/@dcloudio/'
      )
    }
  }

  if (process.env.UNI_PLATFORM === 'mp-alipay') {
    str = str.replace('node-modules/@', 'node-modules/npm-scope-')
  }
  return str
}

export function normalizeMiniProgramFilename(
  filename: string,
  inputDir?: string
) {
  if (!inputDir || !path.isAbsolute(filename)) {
    return normalizeNodeModules(filename)
  }
  return normalizeNodeModules(path.relative(inputDir, filename))
}

export function normalizeParsePlugins(
  importer: string,
  babelParserPlugins?: ParserPlugin[]
) {
  const isTS = EXTNAME_TS_RE.test(importer.split('?')[0])
  const plugins: ParserPlugin[] = []
  if (isTS) {
    plugins.push('jsx')
  }
  if (babelParserPlugins) plugins.push(...babelParserPlugins)
  if (isTS) plugins.push('typescript', 'decorators-legacy')
  return plugins
}

export function pathToGlob(
  pathString: string,
  glob: string,
  options: { windows?: boolean; escape?: boolean } = {}
): string {
  const isWindows =
    'windows' in options ? options.windows : /^win/.test(process.platform)
  const useEscape = options.escape
  const str = isWindows ? pathString.replace(/\\/g, '/') : pathString
  let safeStr = str.replace(
    /[\\*?[\]{}()!]/g,
    isWindows || !useEscape ? '[$&]' : '\\$&'
  )
  return path.posix.join(safeStr, glob)
}

export function resolveSourceMapPath(
  outputDir?: string,
  platform?: UniApp.PLATFORM
) {
  return path.resolve(
    outputDir || process.env.UNI_OUTPUT_DIR,
    '../.sourcemap/' + (platform || getPlatformDir())
  )
}

function hasProjectYarn(cwd: string) {
  return fs.existsSync(path.join(cwd, 'yarn.lock'))
}

function hasProjectPnpm(cwd: string) {
  return fs.existsSync(path.join(cwd, 'pnpm-lock.yaml'))
}

function getInstallCommand(cwd: string) {
  return hasProjectYarn(cwd)
    ? 'yarn add'
    : hasProjectPnpm(cwd)
    ? 'pnpm i'
    : 'npm i'
}

export function installDepTips(
  type: 'dependencies' | 'devDependencies',
  module: string,
  version?: string
) {
  return `Cannot find module: ${module}
Please run \`${colors.cyan(
    `${getInstallCommand(process.cwd())} ${
      module + (version ? '@' + version : '')
    }${type === 'devDependencies' ? ' -D' : ''}`
  )}\` and try again.`
}

/**
 * 根据路径判断是否为 App.(u?)vue
 * @param {string} filename 相对、绝对路径
 * @returns
 */
export function isAppVue(filename: string) {
  const _filePath = normalizePath(filename)
  return /(\/|\\)app\.(u?)vue$/.test(_filePath.toLowerCase())
}

export function resolveAppVue(inputDir: string) {
  if (process.env.UNI_APP_X === 'true') {
    const appUVue = path.resolve(inputDir, 'App.uvue')
    if (fs.existsSync(appUVue)) {
      return normalizePath(appUVue)
    }
  }
  return normalizePath(path.resolve(inputDir, 'App.vue'))
}

export function parseImporter(importer: string) {
  importer = importer.split('?')[0]
  if (path.isAbsolute(importer)) {
    return normalizePath(path.relative(process.env.UNI_INPUT_DIR, importer))
  }
  return importer
}

export function createResolveErrorMsg(source: string, importer: string) {
  return `Could not resolve "${source}" from "${parseImporter(importer)}"`
}

export function enableSourceMap() {
  if (process.env.UNI_APP_SOURCEMAP === 'true') {
    return true
  }
  if (process.env.UNI_APP_SOURCEMAP === 'false') {
    return false
  }
  return process.env.NODE_ENV === 'development' && isNormalCompileTarget()
}

export function requireUniHelpers() {
  require(path.resolve(
    process.env.UNI_HBUILDERX_PLUGINS,
    'uni_helpers/lib/bytenode'
  ))
  return require(path.join(process.env.UNI_HBUILDERX_PLUGINS, 'uni_helpers'))
}

export function normalizeEmitAssetFileName(fileName: string) {
  const extname = path.extname(fileName)

  if (process.env.UNI_APP_X_TSC === 'true') {
    if (extname !== '.ts') {
      return fileName + '.ts'
    }
  } else {
    // logo.png、pages.json 等
    if (!['.ts', '.uts', '.uvue', '.vue'].includes(extname)) {
      fileName = fileName + '.uts'
    }
  }
  return fileName
}

function createIdent(platform: UniApp.PLATFORM) {
  if (process.env.UNI_INPUT_DIR) {
    const manifestJson = parseManifestJsonOnce(process.env.UNI_INPUT_DIR)
    let id = (manifestJson.appid || '').replace('__UNI__', '')
    const platformAppId = manifestJson[platform]?.appid
    if (platformAppId) {
      id += '%%' + platformAppId
    }
    if (id) {
      return Buffer.from(Buffer.from(id).toString('base64')).toString('hex')
    }
  }
  return ''
}

export function createShadowImageUrl(cdn: number, type: string = 'grey') {
  let identStr = ''
  if (process.env.UNI_PLATFORM !== 'h5' && process.env.UNI_PLATFORM !== 'web') {
    const ident = createIdent(process.env.UNI_PLATFORM)
    identStr = ident ? `${ident}/` : ''
  }
  return `https://cdn${
    (cdn || 0) + (process.env.UNI_APP_X === 'true' ? 1000 : 0) || ''
  }.dcloud.net.cn/${identStr}img/shadow-${type}.png`
}

export function isNormalCompileTarget() {
  // 目前有特殊编译目标 uni_modules 和 ext-api
  return !process.env.UNI_COMPILE_TARGET
}

/**
 * copy from entities 用于转义实体字符
 * @param regex
 * @param map
 * @returns
 */
export function getEscaper(
  regex: RegExp,
  map: Map<number, string>
): (data: string) => string {
  return function escape(data: string): string {
    let match
    let lastIndex = 0
    let result = ''

    while ((match = regex.exec(data))) {
      if (lastIndex !== match.index) {
        result += data.substring(lastIndex, match.index)
      }

      // We know that this character will be in the map.
      result += map.get(match[0].charCodeAt(0))!

      // Every match will be of length 1
      lastIndex = match.index + 1
    }

    return result + data.substring(lastIndex)
  }
}
