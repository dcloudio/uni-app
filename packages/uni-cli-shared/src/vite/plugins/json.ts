import path from 'path'
import type { Plugin } from 'vite'
import { parseJson } from '../../json'
import { preJson } from '../../preprocess'
import { parseVueRequest } from '../utils'

export function uniJsonPlugin(): Plugin {
  return {
    name: 'uni:json',
    enforce: 'pre',
    transform(code, id) {
      // 如果已经被其他插件处理过了，就不再处理，比如 commonjs 插件，ICAPRegistrar.json?commonjs-external
      if (id.startsWith('\0')) {
        return
      }
      if (path.extname(parseVueRequest(id).filename) !== '.json') {
        return
      }
      return {
        code: JSON.stringify(parseJson(preJson(code)), null, 2),
        map: {
          mappings: '',
        },
      }
    },
  }
}
