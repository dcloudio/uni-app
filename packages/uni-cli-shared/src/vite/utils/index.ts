import { isNormalCompileTarget } from '../../utils'
import type { ResolvedConfig } from 'vite'

export * from './ast'
export * from './url'
export * from './plugin'
export * from './utils'

// 内置组件css列表，h5平台需要合并进去首页css中
export const buildInCssSet = new Set<string>()

export function isCombineBuiltInCss(config: ResolvedConfig) {
  if (!isNormalCompileTarget()) {
    return false
  }
  return config.command === 'build' && config.build.cssCodeSplit
}
