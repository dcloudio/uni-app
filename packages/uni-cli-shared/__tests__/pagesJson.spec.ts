import path from 'path'
import { checkPagesJson } from '../src/json/utils'
import { generateCodeFrame } from '../src/vite/plugins/vitejs/utils'

describe('pages.json', () => {
  const inputDir = path.resolve(__dirname, './examples/check-pages-json')
  test(`pages check`, () => {
    const source = JSON.stringify(
      {
        pages: [
          {
            path: 'pages/index/index',
          },
          {
            path: 'pages/index/test',
          },
        ],
      },
      null,
      2
    )
    try {
      checkPagesJson(source, inputDir)
    } catch (error: any) {
      expect(error).toMatchSnapshot()
      expect(
        generateCodeFrame(
          source,
          error.offsetStart as number,
          error.offsetEnd as number
        ).replace(/\t/g, ' ')
      ).toMatchSnapshot()
    }
  })
  test(`subPackages check`, () => {
    const source = JSON.stringify(
      {
        subPackages: [
          {
            root: 'pages/API',
            pages: [
              {
                path: 'index/index',
              },
            ],
          },
        ],
      },
      null,
      2
    )
    try {
      checkPagesJson(source, inputDir)
    } catch (error: any) {
      expect(error).toMatchSnapshot()
      expect(
        generateCodeFrame(
          source,
          error.offsetStart as number,
          error.offsetEnd as number
        ).replace(/\t/g, ' ')
      ).toMatchSnapshot()
    }
  })
  test(`subpackages check`, () => {
    const source = JSON.stringify({
      subpackages: [
        {
          root: 'pages/API',
          pages: [
            {
              path: 'index/index',
            },
          ],
        },
      ],
    })
    try {
      checkPagesJson(source, inputDir)
    } catch (error: any) {
      expect(error).toMatchSnapshot()
      expect(
        generateCodeFrame(
          source,
          error.offsetStart as number,
          error.offsetEnd as number
        ).replace(/\t/g, ' ')
      ).toMatchSnapshot()
    }
  })
})
