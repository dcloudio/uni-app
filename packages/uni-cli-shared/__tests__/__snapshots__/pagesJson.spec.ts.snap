// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages.json pages check 1`] = `
{
  "code": "CompilerError",
  "loc": {
    "start": {
      "column": 14,
      "line": 7,
    },
  },
  "message": "The page "pages/index/test" does not exist.",
  "name": "CompilerError",
  "offsetEnd": 100,
  "offsetStart": 82,
}
`;

exports[`pages.json pages check 2`] = `
"5  |      },
6  |      {
7  |        "path": "pages/index/test"
   |                ^^^^^^^^^^^^^^^^^^
8  |      }
9  |    ]"
`;

exports[`pages.json subPackages check 1`] = `
{
  "code": "CompilerError",
  "loc": {
    "start": {
      "column": 18,
      "line": 7,
    },
  },
  "message": "The page "pages/API/index/index" does not exist.",
  "name": "CompilerError",
  "offsetEnd": 112,
  "offsetStart": 99,
}
`;

exports[`pages.json subPackages check 2`] = `
"5  |        "pages": [
6  |          {
7  |            "path": "index/index"
   |                    ^^^^^^^^^^^^^
8  |          }
9  |        ]"
`;

exports[`pages.json subpackages check 1`] = `
{
  "code": "CompilerError",
  "loc": {
    "start": {
      "column": 53,
      "line": 1,
    },
  },
  "message": "The page "pages/API/index/index" does not exist.",
  "name": "CompilerError",
  "offsetEnd": 66,
  "offsetStart": 53,
}
`;

exports[`pages.json subpackages check 2`] = `
"1  |  {"subpackages":[{"root":"pages/API","pages":[{"path":"index/index"}]}]}
   |                                                       ^^^^^^^^^^^^^"
`;
